Composer/packages/client/src/components/ImportModal/ImportModal.tsx

```diff
@@ -0,0 +1,338 @@
+            navigate(`/bot/${existingProject?.id}`);
+          } else {
+            const err = await res.json();
+            throw err;
+          }
+        } catch (e) {
+          console.error('Something went wrong while saving bot to existing project: ', e);
+          setError(e);
+          setModalState('failed');
+        }
+      };
+      backupAndSave();
+    }
+  }, [existingProject, importedProjectInfo]);
+
+  useEffect(() => {
+    if (modalState === 'downloadingContent') {
+      const importBotContent = async () => {
+        if (location && location.href) {
+          try {
+            const { description, name } = importPayload;
+
+            let res = await fetch(
+              `/api/import/${importSource}?payload=${encodeURIComponent(JSON.stringify(importPayload))}`,
+              {
+                method: 'POST',
+              }
+            );
+            if (res.status !== 200) {
+              throw `Something went wrong during import: ${res.status} ${res.statusText}`;
+            }
+            const data = await res.json();
+            const { alias, eTag, templateDir, urlSuffix } = data;
+            const projectInfo = { description, name, templateDir, urlSuffix, eTag, source: importSource, alias };
+            setImportedProjectInfo(projectInfo);
+
+            if (alias) {
+              // check to see if Composer currently has a bot project corresponding to the alias
+              res = await fetch(`/api/projects/alias/${alias}`, { method: 'GET' });
+              if (res.status === 200) {
+                const project = await res.json();
+                setExistingProject(project);
+                // ask user if they want to save to existing, or save as a new project
+                setModalState('promptingToSave');
+                return;
+              }
+            }
+            importAsNewProject(projectInfo);
+          } catch (e) {
+            // something went wrong, abort and navigate to the home page
+            console.error(`Aborting import: ${e}`);
+            navigate('/home');
+          }
+        }
+      };
+      importBotContent();
+    }
+  }, [modalState, importPayload, importSource]);
+
+  useEffect(() => {
+    if (modalState === 'signingIn') {
+      const signIn = async () => {
+        try {
+          const res = await fetch(
+            `/api/import/${importSource}/authenticate?payload=${encodeURIComponent(JSON.stringify(importPayload))}`,
+            {
+              method: 'POST',
+            }
+          );
+          if (res.status !== 200) {
+            throw `Something went wrong during authenticating import: ${res.status} ${res.statusText}`;
+          }
+          setModalState('downloadingContent');
+        } catch (e) {
+          // something went wrong, abort and navigate to the home page
+          console.error(`Aborting import: ${e}`);
+          navigate('/home');
+        }
+      };
+      signIn();
```

Comment:
await this?

Reply:
I was under the impression you couldn't pass an async function to `useCallback()`. I'll change this

Reply:
Is there a pattern for awaiting inside of `useEffect()`? The research I did online pointed to this pattern in which I can't await signIn because it's not inside of an async function.

Reply:
What you are doing is perfectly fine, I was wondering if this component relies on the results of those two async functions since I'm not super familiar with the business logic here.

Another way of achieving the same is using an IIFE:
```
(async () => {
   await awaitable();
})();
```